File: elixir/lib/2021/04_giant_squid/aoc202104.ex
defmodule AOC2021.Day04 do
  @moduledoc """
  Advent of Code 2021, day 4: Giant Squid
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    [moves | boards] = puzzle_input |> String.split("\n\n")

    {moves |> String.split(",") |> Enum.map(&String.to_integer/1),
     boards |> Enum.map(&parse_board/1)}
  end

  def parse_board(board) do
    board
    |> String.split("\n")
    |> Enum.map(fn line ->
      line |> String.split() |> Enum.map(&String.to_integer/1) |> Enum.with_index()
    end)
    |> Enum.with_index()
    |> Enum.reduce(
      %{},
      fn {line, row}, acc ->
        line |> Enum.reduce(acc, fn {number, col}, acc -> Map.put(acc, number, {col, row}) end)
      end
    )
  end

  @doc """
  Solve part 1
  """
  def part1(input) do
    input |> play_bingo_to_first() |> calculate_board_score()
  end

  @doc """
  Solve part 2
  """
  def part2(input) do
    input |> play_bingo_to_last() |> calculate_board_score()
  end

  @doc """
  Play bingo on all boards, return the first winning board and move

  ## Example:

      iex> play_bingo_to_first({[0,1,2,3,4,5,6], [%{0 => {1, 0}, 1 => {0, 1}, 3 => {1, 1}, 5 => {0, 0}}]})
      {3, %{0 => nil, 1 => nil, 3 => nil, 5 => {0, 0}}}
  """
  def play_bingo_to_first({moves, boards}), do: play_bingo_to_first(moves, {boards, nil, nil})

  def play_bingo_to_first([move | moves], {boards, nil, nil}),
    do: play_bingo_to_first(moves, play_move(boards, move))

  def play_bingo_to_first(_moves, {_boards, winning_move, winning_board}),
    do: {winning_move, winning_board}

  @doc """
  Play bingo on all boards, return the last board to win and its winning move

  ## Example:

      iex> play_bingo_to_last(
      ...>   {[0, 1, 2, 3, 4, 5, 6],
      ...>    [
      ...>      %{0 => {1, 0}, 1 => {0, 1}, 3 => {1, 1}, 5 => {0, 0}},
      ...>      %{1 => {1, 0}, 2 => {1, 1}, 3 => {0, 1}, 4 => {0, 0}}
      ...>    ]}
      ...> )
      {3, %{0 => nil, 1 => nil, 3 => nil, 5 => {0, 0}}}
  """
  def play_bingo_to_last({moves, boards}), do: play_bingo_to_last(moves, {boards, nil, nil})

  def play_bingo_to_last(_moves, {[], winning_move, winning_board}),
    do: {winning_move, winning_board}

  def play_bingo_to_last([move | moves], {boards, _, _}),
    do: play_bingo_to_last(moves, play_move(boards, move))

  @doc """
  Play a move on all boards, return information about winning boards

  ## Example:

      iex> play_move(
      ...>   [
      ...>     %{0 => {1, 0}, 1 => {0, 1}, 3 => {1, 1}, 5 => {0, 0}},
      ...>     %{1 => {1, 0}, 2 => {1, 1}, 3 => {0, 1}, 4 => {0, 0}}
      ...>   ],
      ...>   1
      ...> )
      {
        [
          %{0 => {1, 0}, 1 => nil, 3 => {1, 1}, 5 => {0, 0}},
          %{1 => nil, 2 => {1, 1}, 3 => {0, 1}, 4 => {0, 0}}
        ],
        nil,
        nil
      }
  """
  def play_move(boards, move) when is_list(boards) do
    boards |> Enum.map(&play_move(&1, move)) |> check_winning_position(move)
  end

  def play_move(board, move) when is_map(board), do: Map.replace(board, move, nil)

  @doc """
  Check if any board in a list represent a winning position

  ## Examples:

      iex> check_winning_position([%{0 => nil, 1 => {0, 1}, 3 => {1, 1}, 5 => {0, 0}}], 0)
      {[%{0 => nil, 1 => {0, 1}, 3 => {1, 1}, 5 => {0, 0}}], nil, nil}

      iex> check_winning_position([%{0 => nil, 1 => nil, 3 => nil, 5 => {0, 0}}], 3)
      {[], 3, %{0 => nil, 1 => nil, 3 => nil, 5 => {0, 0}}}
  """
  def check_winning_position(boards, move) when is_list(boards) do
    {winning, remaining} = boards |> Enum.split_with(&check_winning_position/1)
    winning_board = winning |> Enum.at(0)

    if winning_board, do: {remaining, move, winning_board}, else: {remaining, nil, nil}
  end

  @doc """
  Check if a board represents a winning position

  ## Examples:

      iex> check_winning_position(%{0 => {1, 0}, 1 => {0, 1}, 3 => {1, 1}, 5 => {0, 0}})
      false

      iex> check_winning_position(%{0 => nil, 1 => nil, 3 => {1, 1}, 5 => {0, 0}})
      false

      iex> check_winning_position(%{0 => nil, 1 => nil, 3 => nil, 5 => {0, 0}})
      true
  """
  def check_winning_position(board) when is_map(board) do
    num_lines = board |> Map.keys() |> length() |> :math.sqrt() |> round()
    all_lines = MapSet.new(0..(num_lines - 1))

    {columns, rows} =
      board
      |> Enum.reject(fn {_, pos} -> pos == nil end)
      |> Enum.into(%{})
      |> Map.values()
      |> Enum.unzip()
      |> then(fn {columns, rows} -> {MapSet.new(columns), MapSet.new(rows)} end)

    if MapSet.equal?(all_lines, columns) and MapSet.equal?(all_lines, rows), do: false, else: true
  end

  @doc """
  Calculate the score of a board position

  ## Example:

      iex> calculate_board_score({3, %{0 => nil, 1 => nil, 3 => nil, 5 => {0, 0}}})
      15
  """
  def calculate_board_score({last_move, board}) do
    board_sum =
      board
      |> Enum.flat_map(fn {number, pos} ->
        case pos do
          nil -> []
          _ -> [number]
        end
      end)
      |> Enum.sum()

    last_move * board_sum
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2021/04_giant_squid/aoc202104.py
"""AoC 4, 2021: Giant Squid."""

# Standard library imports
import math
import pathlib
import sys
from collections import UserList, deque
from dataclasses import dataclass


@dataclass
class BingoBoard:
    """A bingo board including a list of moves."""

    moves: deque[int]
    board: dict[int, tuple[int, int] | None]
    last_move: int | None = None

    @classmethod
    def from_str(cls, moves, board_str):
        """Create a board from a string."""
        board = {}
        for row, line in enumerate(board_str.split("\n")):
            for col, number in enumerate(line.split()):
                board[int(number)] = (row, col)

        return cls(deque(moves), board)

    def check_bingo(self):
        """Check if a board has scored a bingo.

        >>> board = BingoBoard(deque(), {0: (0, 0), 1: None, 2: None, 3: (1,1)})
        >>> board.check_bingo()
        False

        >>> board = BingoBoard(deque(), {0: None, 1: None, 2: (1, 0), 3: (1,1)})
        >>> board.check_bingo()
        True
        """
        num_lines = round(math.sqrt(len(self.board)))
        rows = {pos[0] for pos in self.board.values() if pos is not None}
        cols = {pos[1] for pos in self.board.values() if pos is not None}

        return len(rows) != num_lines or len(cols) != num_lines

    def make_moves(self):
        """Make moves until bingo, will not make additional moves if called a
        second time.

        >>> board = BingoBoard(deque([0, 3, 1, 2]), {0: (0, 0), 1: (0, 1), 2: (1, 0), 3: (1,1)})
        >>> board.make_moves()
        >>> board
        BingoBoard(moves=deque([2]), board={0: None, 1: None, 2: (1, 0), 3: None}, last_move=1)
        """
        while not self.check_bingo():
            move = self.last_move = self.moves.popleft()
            if move in self.board:
                self.board[move] = None

    @property
    def score(self):
        """Score a board.

        >>> board = BingoBoard(moves=deque([2]), board={0: None, 1: None, 2: (1, 0), 3: None}, last_move=1)
        >>> board.score
        2
        """
        board_sum = sum(number for number, pos in self.board.items() if pos is not None)
        return self.last_move * board_sum


class BingoBoards(UserList):
    """List of BingoBoard objects."""

    def play_bingo(self):
        """Play bingo on all boards."""
        for board in self.data:
            board.make_moves()

    @property
    def best_board(self):
        """Find the board that wins first."""
        return max(self.data, key=lambda board: len(board.moves))

    @property
    def worst_board(self):
        """Find the board that wins last."""
        return min(self.data, key=lambda board: len(board.moves))


def parse_data(puzzle_input):
    """Parse input."""
    moves_str, *board_strs = puzzle_input.split("\n\n")
    moves = [int(move) for move in moves_str.split(",")]

    return BingoBoards(
        BingoBoard.from_str(moves, board_str) for board_str in board_strs
    )


def part1(data):
    """Solve part 1."""
    data.play_bingo()
    return data.best_board.score


def part2(data):
    """Solve part 2."""
    data.play_bingo()
    return data.worst_board.score


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
